src/service/url-replacements-impl.js

```diff
@@ -408,55 +489,158 @@ export class UrlReplacements {
+   * version.
    * @param {string} varName
-   * @param {function(*):*} resolver
+   * @param {!SyncResolverDef} syncResolver
    * @return {!UrlReplacements}
    * @private
    */
-  set_(varName, resolver) {
+  set_(varName, syncResolver) {
     dev.assert(varName.indexOf('RETURN') == -1);
-    this.replacements_[varName] = resolver;
+    this.replacements_[varName] = this.replacements_[varName] || {};
+    this.replacements_[varName].sync = syncResolver;
     this.replacementExpr_ = undefined;
     return this;
   }
 
   /**
+  * Sets an async value resolver for the variable with the specified name.
+  * The value resolver may optionally take an extra parameter.
+  * Can be called in conjuction with setAsync to allow for additional
+  * asynchronous resolver where expand will use async and expandSync the sync
+  * version.
+   * @param {string} varName
+   * @param {!AsyncResolverDef} asyncResolver
+   * @return {!UrlReplacements}
+   * @private
+   */
+  setAsync_(varName, asyncResolver) {
+    dev.assert(varName.indexOf('RETURN') == -1);
+    this.replacements_[varName] = this.replacements_[varName] || {};
+    this.replacements_[varName].async = asyncResolver;
+    this.replacementExpr_ = undefined;
+    return this;
+  }
+
+  /**
+   * Utility function for setting resolver for timing data that supports
+   * sync and async.
+   * @param {string} varName
+   * @param {string} startEvent
+   * @param {string=} endEvent
+   * @return {!UrlReplacements}
+   * @private
+   */
+  setTimingResolver_(varName, startEvent, endEvent) {
+    return this.set_(varName, () => {
+        return this.getTimingDataSync_(startEvent, endEvent);
+      }).setAsync_(varName, () => {
+        return this.getTimingDataAsync_(startEvent, endEvent);
+      });
+  }
+
+  /**
+   * Calling encodeURIComponent with non-string type fails type check despite
+   * browser supporting it (simply converts to string).  This allows for being
+   * explicit about the behavior.
+   * @param {*} val
+   * @return {!string}
+   * @private
+   */
+  encodeValue_(val) {
+    switch (typeof val) {
+      case 'string':
+        return encodeURIComponent(val);
+      case 'number':
+      case 'boolean':
+        return String(val);
+      default:  // null evaluates as object
+        return '';
+    }
+  }
+
+/**
+ * Synchronously expands the provided URL by replacing all known variables with
+ * their resolved values. Optional `opt_bindings` can be used to add new
+ * variables or override existing ones.  Any async bindings are ignored.
+ * @param {string} url
+ * @param {!Object<string, (ResolverReturnDef|!SyncResolverDef)>=} opt_bindings
+ * @param {!Object<string, ResolverReturnDef>=} opt_collectVars
+ * @return {string}
+ * @private
+ */
+expandSync(url, opt_bindings, opt_collectVars) {
+  if (!this.initialized_) {
+    this.initialize_();
+  }
+  const expr = this.getExpr_(opt_bindings);
+  url = url.replace(expr, (match, name, opt_strargs) => {
```

Comment:
expandSync only allows replacement keys with synchronous resolvers.  Document click listener sets the new href value synchronously so that it is used by navigation.  From discussions with @cramforce it was decided to not support async replacement to ensure navigation is not delayed.